File: 200.number-of-islands.js
/*
 * @lc app=leetcode id=200 lang=javascript
 *
 * [200] Number of Islands
 */

// @lc code=start
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
  // Iterate over all elements in grid
  // BFS -> iterative move UDLR
  // base case:
  // 1. Next element UDLR is 0
  // 2. Or element is undefined (edge)
  // Prevent double counting
  // Use 2D array of true/false to keep track of visited
  // O((m * n)) time
  // O((m * n)) space
  let m = grid.length
  let n = grid[0].length
  let counter = 0
  let hasVisitedCache = Array(m).fill().map(() => new Array(n).fill(false))
  // iterate over all elements in grid
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      // BFS -> iterative move UDLR
      if (grid[i][j] === '0' ||
          hasVisitedCache[i][j]) {
        hasVisitedCache[i][j] = true
        continue
      }
      counter++
      let queue = []
      // [row, col]
      queue.unshift([i, j])
      while (queue.length > 0) {
        let cordinates = queue.shift()
        let row = cordinates[0]
        let col = cordinates[1]
        if (hasVisitedCache[row][col]) continue
        hasVisitedCache[row][col] = true
        let up = row + 1
        let down = row - 1
        let left = col - 1
        let right = col + 1
        // check up 
        if (grid[up] && grid[row][col] !== undefined &&
            grid[up][col] === '1') {
              queue.unshift([up, col])
            }
        // check down
        if (grid[down] && grid[down][col] !== undefined &&
            grid[down][col] === '1') {
              queue.unshift([down, col])
            }
        // check left
        if (grid[row][left] !== undefined &&
            grid[row][left] === '1') {
              queue.unshift([row, left])
            }
        // check right
        if (grid[row][right] !== undefined &&
            grid[row][right] === '1') {
              queue.unshift([row, right])
            }
      }
    }
  }
  return counter
};
// @lc code=end

File: cpp/200.number-of-islands.cpp
/*
 * @lc app=leetcode id=200 lang=cpp
 *
 * [200] Number of Islands
 */

// @lc code=start
class Solution
{
public:
    void dfsHelper(vector<vector<char>> &grid, int row, int col)
    {
        // base cases
        // out of bounds
        if ((row > grid.size() - 1) || row < 0)
        {
            return;
        }
        if ((col > grid[0].size() - 1) || col < 0)
        {
            return;
        }
        // mark current position
        grid[row][col] = 'X';
        // check adjacent positions for '1'
        int up = row - 1;
        int down = row + 1;
        int left = col - 1;
        int right = col + 1;
        // check up
        if (up > -1)
        {
            if (grid[up][col] == '1')
            {
                dfsHelper(grid, up, col);
            }
        }
        // check down
        if (down < grid.size())
        {
            if (grid[down][col] == '1')
            {
                dfsHelper(grid, down, col);
            }
        }
        // check left
        if (left > -1)
        {
            if (grid[row][left] == '1')
            {
                dfsHelper(grid, row, left);
            }
        }
        // check right
        if (right < grid[0].size())
        {
            if (grid[row][right] == '1')
            {
                dfsHelper(grid, row, right);
            }
        }
        return;
    }
    int numIslands(vector<vector<char>> &grid)
    {
        // iterate over entire grid
        // launch a DFS at every position
        // where there is a 1, and has not been visited
        // mark positions visited with '-1'
        // base cases
        // 1. No adjacent '1'
        // 2. reach edge of grid
        // O(m*n) time O(m*n) space
        int numOfIslands = 0;
        for (int row = 0; row < grid.size(); row++)
        {
            for (int col = 0; col < grid[0].size(); col++)
            {
                if (grid[row][col] == '1')
                {
                    // launch DFS to mark entire island
                    dfsHelper(grid, row, col);
                    numOfIslands++;
                }
            }
        }
        return numOfIslands;
    }
};
// @lc code=end
